Sources/MIDIKit/MIDINetwork.swift

```diff
@@ -0,0 +1,199 @@
+        if let endpoint = networkSession?.destinationEndpoint() {
+            return MIDIEndpoint(endpoint)
+        }
+        return nil
+    }
+    
+    public func removeAllNetworkConnections() {
+        for connection in networkSession?.connections() ?? [] {
+            networkSession?.removeConnection(connection)
+        }
+    }
+    
+    public private(set) var lastKnownConnection: MIDINetworkConnection? {
+        get {
+            let ud = UserDefaults.standard
+            
+            if let name = ud.string(forKey: "MIDIKit-networkConnection-name"),
+               let hostName  = ud.string(forKey: "MIDIKit-networkConnection-host-name") {
+                
+                let port = ud.integer(forKey: "MIDIKit-networkConnection-host-port")
+                
+                return MIDINetworkConnection(host: .init(name: name, address: hostName, port: port))
+            }
+            
+            return nil
+        }
+        set {
+            let ud = UserDefaults.standard
+            if let connection = newValue {
+                ud.setValue(connection.host.name, forKey: "MIDIKit-networkConnection-name")
+                ud.setValue(connection.host.address, forKey: "MIDIKit-networkConnection-host-name")
+                ud.setValue(connection.host.port, forKey: "MIDIKit-networkConnection-host-port")
+            } else {
+                ud.removeObject(forKey: "MIDIKit-networkConnection-name")
+                ud.removeObject(forKey: "MIDIKit-networkConnection-host-name")
+                ud.removeObject(forKey: "MIDIKit-networkConnection-host-port")
+            }
+            ud.synchronize()
+        }
+    }
+}
+
+public class BonjourService: NSObject, NetServiceBrowserDelegate, NetServiceDelegate {
+    
+    var timeout: TimeInterval = 1.0
+    var serviceFoundClosure: (([NetService]) -> Void)!
+    var domainFoundClosure: (([String]) -> Void)!
+
+    public struct Services {
+        public static let Apple_Midi: String = "_apple-midi._udp."
+    }
+    public static let LocalDomain: String = "local."
+
+    public let serviceBrowser: NetServiceBrowser = NetServiceBrowser()
+    public var services = [NetService]()
+    public var domains = [String]()
+    public var isSearching: Bool = false
+    public var serviceTimeout: Timer = Timer()
+    public var domainTimeout: Timer = Timer()
+    
+    var servicesToResolve = 0
+
+    @discardableResult
+    public func findService(_ identifier: String, domain: String, found: @escaping ([NetService]) -> Void) -> Bool {
+        if !isSearching {
+            serviceBrowser.delegate = self
+            serviceTimeout = Timer.scheduledTimer(
+                timeInterval: self.timeout,
+                target: self,
+                selector: #selector(BonjourService.noServicesFound),
+                userInfo: nil,
+                repeats: false)
+            serviceBrowser.searchForServices(ofType: identifier, inDomain: domain)
```

Comment:
`Timer.scheduledTimer(timeInterval:target:selector:userInfo:repeats:)` and `serviceBrowser.searchForServices(ofType:inDomain:)` introduce concurrency.
`Timer.scheduledTimer(timeInterval:target:selector:userInfo:repeats:)`will call `BonjourService.noServicesFound()` on the current run loop.
AFAIK `NetServiceBrowser` gives no guarantee about the Thread/RunLoop/DispatchQueue which it calls its delegate.
This results in a race condition because both, the timeout and the service browser, access `self`.

Reply:
To be quite honest I’m not sure the service discovery has to be part of the MIDI client. Better to leave this task to the developer, using the preferred Bonjour implementation. I’ll scrap it from MIDIKit.